controller/api/destination/endpoint_translator.go

```diff
@@ -287,6 +307,154 @@ func (et *endpointTranslator) sendClientAdd(set watcher.AddressSet) {
 	}
 }
 
+// watchEndpointPolicy establishes a watch on the the policy server for each
+// pod address.
+//
+// The policy server client will error if the stream is closed or the pod
+// address is removed. When the pod address is removed, closeEndpointPolicy
+// will close the client.
+//
+// The policy server client handles Recv errors and signals the updates
+// routine to shutdown.
+func (et *endpointTranslator) watchEndpointPolicy(set watcher.AddressSet) {
+	// When adding a new client to the policy watches map, ensure clients are
+	// not also being deleted by closeEndpointPolicy
+	et.policyWatches.mutex.Lock()
+	defer et.policyWatches.mutex.Unlock()
+
+	for _, addr := range set.Addresses {
+		// If the address is not backed by a pod then there will be no policy
+		// server to watch.
+		if addr.Pod == nil {
+			continue
+		}
+
+		opaquePorts, ok, err := getPodOpaquePortsAnnotations(addr.Pod)
+		if err != nil {
+			et.log.Errorf("failed to get opaque ports annotation for pod: %s", err)
+		}
+
+		// If the opaque ports annotation was not set, then use the
+		// endpoint translator's default value.
+		if !ok {
+			opaquePorts = et.defaultOpaquePorts
+		}
+
+		skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(addr.Pod)
+		if err != nil {
+			et.log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
+		}
+
+		// Create a client which watches for changes to the inbound policy for
+		// portSpec.
+		ctx, cancel := context.WithCancel(et.stream.Context())
+		policyPortSpec := &policyPb.PortSpec{
+			Workload: fmt.Sprintf("%s:%s", addr.Pod.Namespace, addr.Pod.Name),
+			Port:     addr.Port,
+		}
+		client, err := et.policyClient.WatchPort(ctx, policyPortSpec)
+		if err != nil {
+			et.log.Errorf("failed to create policy server watch for %s:%d: %s", policyPortSpec.Workload, policyPortSpec.Port, err)
+		}
+		et.log.Debugf("Establishing watch on policy server %s:%d", policyPortSpec.Workload, policyPortSpec.Port)
+
+		// policyPb.PortSpec cannot be used as a map key, so we create an
+		// equivalent port spec with comparable keys.
+		portSpec := &portSpec{
+			workload: policyPortSpec.Workload,
+			port:     policyPortSpec.Port,
+		}
+		portClient := watchPortClient{
+			client: client,
+			cancel: cancel,
+		}
+
+		// Before adding the new port client, ensure any previous client for
+		// the port spec has been closed.
+		if pc, ok := et.policyWatches.watches[*portSpec]; ok {
+			pc.cancel()
+		}
+
+		et.policyWatches.watches[*portSpec] = &portClient
```

Comment:
What's the intention with this cache of watches?  Right now if two different requests come in which include the same pod+port then the later policy watch will replace the earlier one in this map and the earlier one will no longer be cancelable.

Is the goal to share the same policy watch between both requests?  If so, how do updates from the policy watch get fanned out to both destination requests?